src/fetch_orderbook.ts

```diff
@@ -0,0 +1,86 @@
+import Web3 from "web3";
+import { AbiItem } from "web3-utils";
+import { decodeOrdersBN, OrderBN } from "@gnosis.pm/dex-contracts/src/index";
+import { Orderbook, Offer } from "@gnosis.pm/dex-contracts/src/orderbook";
+import { Fraction } from "@gnosis.pm/dex-contracts/src/fraction";
+import { BatchExchangeViewer } from "@gnosis.pm/dex-contracts/build/types/BatchExchangeViewer";
+import batchExchangeViewerAbi from "@gnosis.pm/dex-contracts/build/contracts/BatchExchangeViewer.json";
+
+let lastBatchId = 0;
+let lastOrderbooks: Map<string, Orderbook> = null;
+
+const web3 = new Web3("https://node.mainnet.gnosisdev.com");
+const batchExchangeViewer = new web3.eth.Contract(
+  batchExchangeViewerAbi.abi as AbiItem[],
+  batchExchangeViewerAbi.networks[1].address
+) as BatchExchangeViewer;
+
+export async function fetchOrderbook() {
+  const block = await web3.eth.getBlock("latest");
+  const currentBatchId = batchIdFromTimestamp(block.timestamp as number);
+  if (currentBatchId > lastBatchId) {
+    lastBatchId = currentBatchId;
+    // tslint:disable-next-line:no-console
+    console.log(`Fetch orderbook for batch: ${currentBatchId}`);
+    const orders = await getOpenOrdersPaginated(batchExchangeViewer, 50);
+    lastOrderbooks = new Map();
+    orders.forEach(item => {
+      addItemToOrderbooks(lastOrderbooks, item);
+    });
+    // tslint:disable-next-line:no-console
+    console.log(`Orderbook fetched for batch: ${currentBatchId}`);
+  }
+  return lastOrderbooks;
+}
+
+function batchIdFromTimestamp(timestamp: number) {
+  return Math.floor(timestamp / 300);
+}
+
+async function getOpenOrdersPaginated(
+  instance: BatchExchangeViewer,
+  pageSize: number
+) {
+  let orders: OrderBN[] = [];
+  let nextPageUser = "0x0000000000000000000000000000000000000000";
+  let nextPageUserOffset = "0";
+  let lastPageSize = pageSize;
+
+  while (lastPageSize === pageSize) {
+    // tslint:disable-next-line:no-console
+    console.log("Fetching Page");
+    const page = await instance.methods
+      .getOpenOrderBookPaginated([], nextPageUser, nextPageUserOffset, pageSize)
+      .call();
+    const elements = decodeOrdersBN(page.elements);
+    orders = orders.concat(elements);
+
+    // Update page info
+    lastPageSize = elements.length;
+    nextPageUser = page.nextPageUser;
+    nextPageUserOffset = page.nextPageUserOffset;
+  }
+  return orders;
+}
+
+function addItemToOrderbooks(
+  orderbooks: Map<string, Orderbook>,
+  item: OrderBN
+) {
+  let orderbook = new Orderbook(item.sellToken, item.buyToken);
+  if (!orderbooks.has(orderbook.pair())) {
+    orderbooks.set(orderbook.pair(), orderbook);
+  }
+  orderbook = orderbooks.get(orderbook.pair());
```

Comment:
ubernit: Maybe calculating a pair ID from buy and sell tokens can be factored out. I think that would make this snippet a little cleaner.